### ISSUE TITLE
Postgres proxy: Envoy Crash during in pgbench test

### ISSUE THREAD

**Issue by shiponcs** [envoyproxy/envoy#36471]
*Title*: *Envoy Crash during in pgbench test*

*Description*:
We have run pgbench to test postgres server proxied by Envoy. During the test Envoy crashed with the following log:

```bash
[2024-09-19 14:22:51.227][42][critical][backtrace] [./source/server/backtrace.h:127] Caught Segmentation fault, suspect faulting address 0x0
[2024-09-19 14:22:51.227][42][critical][backtrace] [./source/server/backtrace.h:111] Backtrace (use tools/stack_decode.py to get line numbers):
[2024-09-19 14:22:51.227][42][critical][backtrace] [./source/server/backtrace.h:112] Envoy version: ca7ce529646b1f7c8c760e0e249e285cee4549ed/1.30.2/Modified/RELEASE/BoringSSL
[2024-09-19 14:22:51.227][42][critical][backtrace] [./source/server/backtrace.h:114] Address mapping: 560479564000-56047cf63000 /usr/local/bin/envoy
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #0: [0x7fa621681520]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #1: [0x56047b34bfc4]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #2: [0x56047b34949a]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #3: [0x56047b348d88]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #4: [0x56047b345b6d]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #5: [0x56047b3304fa]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #6: [0x56047b32fd00]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #7: [0x56047b3443e7]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #8: [0x56047bb7aec5]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #9: [0x56047bb72663]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #10: [0x56047bb6f2f8]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #11: [0x56047bb4d2f1]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #12: [0x56047bb4e89d]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #13: [0x56047ca79480]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #14: [0x56047ca77dc1]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #15: [0x56047ad6332f]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #16: [0x56047cafac63]
[2024-09-19 14:22:51.228][42][critical][backtrace] [./source/server/backtrace.h:121] #17: [0x7fa6216d3ac3]
ConnectionImpl 0x26537f7fe640, connecting_: 0, bind_error_: 0, state(): Open, read_buffer_limit_: 32768
socket_: 
  ListenSocketImpl 0x26537f0aeab0, transport_protocol_: raw_buffer
  connection_info_provider_: 
    ConnectionInfoSetterImpl 0x26537f09e518, remote_address_: 100.126.0.213:54396, direct_remote_address_: 100.126.0.213:54396, local_address_: 100.126.5.153:50815, server_name_:
```
*Testing scope:*
Envoy configuration: https://gist.github.com/shiponcs/600fff61cddba5911bc3b9e538a4fc8c
pgbench command: `pgbench  -h localhost -p 50815 -U postgres -c 50 -j 1 -T 00 example`

*Relevant*:
We have tried to find the reason; by running git bisect we discovered that, [this commit](https://github.com/envoyproxy/envoy/commit/61c9a47b8f68c4e66b584fbe86b046e082fec2d9) is causing the crash. If we get rid of the changes introduced in the commit, the issue gets resolved.

**Comment by alyssawilk** [comment]
Thanks for running this by Envoy security before (with approval) posting in the clear

@fabriziomello @cpakulski can you take a look?   I don't think we should necessarily roll back an Envoy build config change based on a broken contrib extension

**Comment by cpakulski** [comment]
Thanks for reporting it. I will try to repro the crash to find the cause. 

>  I don't think we should necessarily roll back an Envoy build config change based on a broken contrib extension.

I agree. I think that the build config probably just exposes the problem but is not the cause of the crash.

**Comment by fabriziomello** [comment]
@cpakulski Long time since I ran this regression tests https://github.com/fabriziomello/envoy-postgres-regression. It requires some reworking to work properly again and don't need to waste resources building Envoy. 

One naive question, those -dev docker images are a kind of nightly-builds?

**Comment by cpakulski** [comment]
@fabriziomello I think that this is what you are looking for: https://hub.docker.com/layers/envoyproxy/envoy/contrib-debug-dev/images/sha256-770fe2414e700156673389677dedca96133fb7b69cb2d9a222253e15fd25b11f?context=explore

**Comment by cpakulski** [comment]
I am still planning to investigate the cause.

**Comment by cpakulski** [comment]
It is under investigation. I am building test bed.

**Comment by caoyukun0430** [comment]
> It is under investigation. I am building test bed.

Hi @cpakulski  thank you very much for taking the effort to debug it! Have you managed to reproduce the issue? Please let us know if anything useful is needed, e.g. we can try to reproduce from our side and provide the proper backtrace and coredump if they can help to speed up the debugging.

**Comment by cpakulski** [comment]
wip.

**Comment by shiponcs** [comment]
Hi @cpakulski,
Are you planning on replacing/fixing the external sql parsing library that's causing the crash? I have found the similar crash recently that we can prevent by just disabling sql parsing.

**Comment by cpakulski** [comment]
Yes, I still plan to work on it and fix the issue. Replacement of the parser library is not considered right now, as it would require rather large amount of work.

**Comment by shiponcs** [comment]
Is this related to this- https://github.com/envoyproxy/envoy/blob/main/contrib/common/sqlutils/source/BUILD#L17 ?

**Comment by Licht-T** [comment]
@cpakulski I just looked into this - `-fsized-deallocation` only **unmasks** the bug. The real fault is in [`envoyproxy/sql-parser`](https://github.com/envoyproxy/sql-parser) (pinned at [`3b40ba2d`](https://github.com/envoyproxy/sql-parser/tree/3b40ba2d106587bdf053a292f7e3bb17e818a57f) in `bazel/repository_locations.bzl`):

1. [`DELTA`](https://github.com/envoyproxy/sql-parser/blob/3b40ba2d106587bdf053a292f7e3bb17e818a57f/src/parser/bison_parser.y#L174) is declared as a reserved token but no grammar rule references it (also dead in [`sql_keywords.txt#L95`](https://github.com/envoyproxy/sql-parser/blob/3b40ba2d106587bdf053a292f7e3bb17e818a57f/src/parser/sql_keywords.txt#L95) and [`flex_lexer.l#L128`](https://github.com/envoyproxy/sql-parser/blob/3b40ba2d106587bdf053a292f7e3bb17e818a57f/src/parser/flex_lexer.l#L128)), so `INSERT INTO pgbench_history (..., delta) VALUES (...)` fails to parse.
1. On every such failure, [the `<str_vec>` `%destructor`](https://github.com/envoyproxy/sql-parser/blob/3b40ba2d106587bdf053a292f7e3bb17e818a57f/src/parser/bison_parser.y#L147-L154) calls `delete` on `char*` entries that the lexer allocated via [`strdup()`](https://github.com/envoyproxy/sql-parser/blob/3b40ba2d106587bdf053a292f7e3bb17e818a57f/src/parser/flex_lexer.l#L251) - `malloc`/`delete` mismatch (UB). Under tcmalloc, sized deallocation returns those chunks to the wrong-sized freelist, and after enough mismatches an unrelated allocation segfaults.

This destructor bug was already fixed upstream in [`hyrise/sql-parser` PR #221](https://github.com/hyrise/sql-parser/pull/221) (issue [#220](https://github.com/hyrise/sql-parser/issues/220)). **Please note that this is not limited to the `DELTA` keyword but the generic parse error, and the same issue is on mysql_proxy as well.**

ASAN also confirms `alloc-dealloc-mismatch (malloc vs operator delete)` at `yydestruct`, allocated by `__interceptor_strdup`.

**Minimal repro:** With `enable_sql_parsing: true` and `PGSSLMODE=disable`:

```sh
echo "INSERT INTO pgbench_history (tid, bid, aid, delta) VALUES (1,1,1,1);" > q.sql
PGSSLMODE=disable PGPASSWORD=postgres pgbench -h envoy -p 50815 -U postgres -c 1 -j 1 -t 500 -f q.sql example
```

Crashes in ~10 iterations on 1.30.2, ~1,600 on `1.36.0-dev` - same bug, tcmalloc layout differences shift when the corruption surfaces.

I will create a PR for this.
